tests/common/exceptions.h

```diff
@@ -0,0 +1,302 @@
+    // Uses r-value reference directly
+    out << "with " << description << ": '" << value << "'\n";
+  };
+  auto append_cstr = [&append_str](const char* description, const char* value) {
+    if (!value) {
+      value = "nullptr";
+    }
+    append_str(description, {value});
+  };
+
+  // Collect exception details, considering possible implementation gaps
+  if constexpr (!detail::has_member::category<sycl::exception>::value) {
+    append_cstr("category", "not supported by implementation");
+  } else {
+    // Using reference to avoid object slicing
+    const auto& category = e.category();
+    append_cstr("category name", category.name());
+  }
+
+  if constexpr (!detail::has_member::code<sycl::exception>::value) {
+    append_cstr("code", "not supported by implementation");
+  } else {
+    // Using reference to avoid object slicing
+    const auto& code = e.code();
+    append_str("code value", std::to_string(code.value()));
+    append_str("code message", code.message());
+  }
+
+  append_cstr("what", e.what());
+
+  return out.str();
+}
+
+/**
+ *  Matchers' implementation details
+ */
+namespace detail {
+
+/**
+ *  Matcher for std::error_category value, with no error code comparison
+ */
+class matcher_exception_category : public Catch::Matchers::MatcherGenericBase {
+ public:
+  matcher_exception_category(const std::error_category& category)
+      : m_category(category) {}
+
+  static constexpr bool is_supported_by_implementation() {
+    return detail::has_member::category<sycl::exception>::value;
+  }
+
+  bool match(const sycl::exception& other) const {
+    if constexpr (!is_supported_by_implementation()) {
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      return other.category() == m_category;
+    }
+  }
+
+  std::string describe() const override {
+    return std::string("has category: '") + m_category.name() + "'";
+  }
+
+ private:
+  const std::error_category& m_category;
+};
+
+/**
+ *  Matcher for sycl::errc error codes
+ *  C++ provides semantic match for std::error_code by operator==
+ */
+struct matcher_exception_code_semantic_match
+    : public Catch::Matchers::MatcherGenericBase {
+  matcher_exception_code_semantic_match(const sycl::errc& code_value)
+      : m_code_value(code_value) {}
+
+  static constexpr bool is_supported_by_implementation() {
+    return detail::has_member::code<sycl::exception>::value;
+  }
+
+  bool match(const sycl::exception& other) const {
+    if constexpr (!is_supported_by_implementation()) {
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      // Semantic match
+      return other.code() == m_code_value;
+    }
+  }
+
+  std::string describe() const override {
+    std::ostringstream result;
+    result << "has code semantically matching to ";
```

Comment:
I don't understand what you mean by "semantic match" here (and above), why not just "match"?

Reply:
My bad. I've got [p0824r1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0824r1.html) in mind and forgot that now SYCL simply has separate error code enums for each category, with no generic errors due to `error_condition` removal. Will fix and rename the matcher.

Reply:
As result, there are 3 matchers now:

- `has_exception_category` - compares category only
- `equals_exception` - for generic errors, with `sycl::errrc` error codes and `sycl::sycl_category()` category
- `equals_exception_for` - for backend-specific errors

Though the `has_exception_category` has a limited functionality now, it was left to have all possible use-cases covered